internal/searchindex/searchindex.go

```diff
@@ -0,0 +1,463 @@
+// Package searchindex contains internal representation of the Atlas SearchIndex resource
+package searchindex
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/cmp"
+
+	"go.mongodb.org/atlas-sdk/v20231115008/admin"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/compat"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/pointer"
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1/common"
+)
+
+// SearchIndex is the internal representation of the Atlas SearchIndex resource for the AKO usage
+// SearchIndexes represented differently in AKO as CRDs and in Atlas as atlas internal structures.
+// Having a separate representation allows for simpler testing of the internal logic, not tied to
+// AKO and Atlas structures
+type SearchIndex struct {
+	akov2.SearchIndex
+	akov2.AtlasSearchIndexConfigSpec
+	ID     *string
+	Status *string
+}
+
+func (s *SearchIndex) GetID() string {
+	return pointer.GetOrDefault(s.ID, "")
+}
+
+func (s *SearchIndex) GetStatus() string {
+	return pointer.GetOrDefault(s.Status, "")
+}
+
+// NewSearchIndexFromAKO requires both parts of search index: data-related part from a Deployment, and
+// index configuration represented in a separate CRD. Partial construction is disabled as it won't work
+// for comparing indexes between each other.
+func NewSearchIndexFromAKO(index *akov2.SearchIndex, config *akov2.AtlasSearchIndexConfigSpec) *SearchIndex {
+	if index == nil || config == nil {
+		return nil
+	}
+	return &SearchIndex{
+		SearchIndex:                *index.DeepCopy(),
+		AtlasSearchIndexConfigSpec: *config.DeepCopy(),
+	}
+}
+
+// NewSearchIndexFromAtlas returns internal representation of the SearchIndex converted from Atlas
+// internals. It can return an error in case some fields are not valid JSON.
+func NewSearchIndexFromAtlas(index admin.ClusterSearchIndex) (*SearchIndex, error) {
```

Comment:
This method has little unit test coverage, please increase as it is central. We just found a bug in here as well.